crates/sui-core/src/accumulators/mod.rs

```diff
@@ -0,0 +1,248 @@
+        .epoch_start_config()
+        .accumulator_root_obj_initial_shared_version()
+    else {
+        debug!("accumulator root object does not exist, skipping accumulator update");
+        return vec![];
+    };
+
+    struct Update {
+        merge: MergedValueIntermediate,
+        split: MergedValueIntermediate,
+    }
+
+    let mut updates = HashMap::<_, Update>::new();
+
+    let mut addresses = HashMap::<_, AccumulatorAddress>::new();
+
+    for effect in ckpt_effects {
+        let tx = effect.transaction_digest();
+        // TransactionEffectsAPI::accumulator_events() uses a linear scan of all
+        // object changes and allocates a new vector. In the common case (on validators),
+        // we still have still have the original vector in the writeback cache, so
+        // we can avoid the unnecessary work by just taking it from the cache.
+        let events = match cache.and_then(|c| c.take_accumulator_events(tx)) {
+            Some(events) => events,
+            None => effect.accumulator_events(),
+        };
+
+        for AccumulatorEvent {
+            accumulator_obj,
+            write:
+                AccumulatorWriteV1 {
+                    operation,
+                    value,
+                    address,
+                },
+        } in events
+        {
+            if let Some(prev) = addresses.insert(accumulator_obj, address.clone()) {
+                debug_assert_eq!(prev, address);
+            }
+
+            let entry = updates.entry(accumulator_obj).or_insert_with(|| {
+                let zero = MergedValueIntermediate::zero(&value);
+                Update {
+                    merge: zero,
+                    split: zero,
+                }
+            });
+
+            match operation {
+                AccumulatorOperation::Merge => {
+                    entry.merge.accumulate_into(value);
+                }
+                AccumulatorOperation::Split => {
+                    entry.split.accumulate_into(value);
+                }
+            }
+        }
+    }
+
+    let mut builder = ProgrammableTransactionBuilder::new();
+
+    let root = builder
+        .input(CallArg::Object(ObjectArg::SharedObject {
+            id: SUI_ACCUMULATOR_ROOT_OBJECT_ID,
+            initial_shared_version: accumulator_root_obj_initial_shared_version,
+            mutable: true,
+        }))
+        .unwrap();
+
+    for (accumulator_obj, update) in updates {
+        let Update { merge, split } = update;
+        let address = addresses.get(&accumulator_obj).unwrap();
+        let merged_value = MergedValue::from(merge);
+        let split_value = MergedValue::from(split);
+        MergedValue::add_move_call(merged_value, split_value, root, address, &mut builder);
+    }
+
+    let epoch_arg = builder.pure(epoch).unwrap();
+    let checkpoint_height_arg = builder.pure(checkpoint_height).unwrap();
+    let idx_arg = builder.pure(0u64).unwrap();
```

Comment:
If you put an index here, assuming it will eventually be used for the purpose of having multiple settlements, you probably also need a total_count argument.

Reply:
why? we just need to make sure each settlement tx has a unique digest. total_count isn't necessary for that, is it?

Reply:
I was thinking that this might be needed to implement parallel execution of multiple settlements. But I guess we could add it then if we need it.